File: c_datastruct/链表/01_环形链表.c
#include <stdio.h>
#include <stdbool.h>

struct ListNode{
    int val;
    struct ListNode* next;
};
bool hasCycle(struct ListNode *head) {
    if(head == NULL || head -> next == NULL)
        return false;
    struct  ListNode* slow = head->next;
    struct  ListNode* fast = head->next->next;
    //快慢指针的思想
    while (fast != slow){
        if(fast == NULL || fast->next == NULL)
            return false;
        slow = slow->next;
        fast = fast->next->next;
    }
    return true;
}

File: java_datastruct/src/main/java/com/meicilly/链表/_01_环形链表.java
package com.meicilly.链表;

public class _01_环形链表 {
    public class ListNode {
        int val;
        ListNode next;
        ListNode(int x) {
            val = x;
        }
    }
    public boolean hasCycle(ListNode head){
        if(head == null || head.next == null) return  false;
        ListNode slow = head.next;
        ListNode fast = head.next.next;
        while (slow != fast){
            if(fast == null || fast.next == null) return false;
            slow = slow.next;
            fast = fast.next.next;
        }
        return true;
    }
}
